syntax = "proto3";

package pl.vizier.messages;

import "github.com/gogo/protobuf/gogoproto/gogo.proto";
import "src/carnot/planpb/plan.proto";
import "src/common/uuid/proto/uuid.proto";
import "src/table_store/proto/schema.proto";
import "src/shared/k8s/metadatapb/metadata.proto";

option go_package = "pixielabs.ai/pixielabs/src/vizier/messages/messagespb;messages";

// A wrapper around all messages which can be sent over the message bus.
message VizierMessage {
    oneof msg {
        RegisterAgentRequest register_agent_request = 1;
        RegisterAgentResponse register_agent_response = 2;
        UpdateAgentRequest update_agent_request = 3;
        UpdateAgentResponse update_agent_response = 4;
        Heartbeat heartbeat = 5;
        HeartbeatAck heartbeat_ack = 6;
        HeartbeatNack heartbeat_nack = 7;
        ExecuteQueryRequest execute_query_request = 8;
    }
}

// AgentInfo contains information about host and agent running on a given machine.
message AgentInfo {
    uuidpb.UUID agent_id      = 1 [
                                  (gogoproto.customname) = "AgentID"
                                  ];
    HostInfo host_info = 2;
}

// HostInfo contains the details for the host (OS, kernel, CPU, etc).
message HostInfo {
    string hostname = 1;
}

message RegisterAgentRequest {
    AgentInfo info = 1;
    AgentUpdateInfo update_info = 2;
}

message RegisterAgentResponse {}

message UpdateAgentRequest {
    AgentInfo info = 1;
    AgentUpdateInfo update_info = 2;
}

message UpdateAgentResponse {}

message AgentUpdateInfo {
    repeated pl.shared.k8s.metadatapb.ContainerInfo containers = 1;
    repeated pl.table_store.schemapb.Relation relations = 2;
}

// The heart beat signal.
message Heartbeat {
    uuidpb.UUID agent_id = 1 [
      (gogoproto.customname) = "AgentID"
                             ];
    // The current unix time in ns on the machine.
    int64 time = 2;
    AgentUpdateInfo update_info = 3;
    int64 sequence_number = 4;
}

enum MetadataResourceType {
    METADATA_RESOURCE_TYPE_UNKNOWN = 0;
    POD = 1;
    SERVICE = 2;
}

message MetadataUpdateInfo {
    message ResourceUpdate {
        // The resource's unique identifier.
        string uid = 1;

        // The resource's name.
        string name = 2;

        // The type of the resource.
        MetadataResourceType type = 3;
    }
    repeated ResourceUpdate updates = 4;
}

// Response sent for a successful heart beat.
message HeartbeatAck {
    // The current unix time in ns on the machine.
    int64 time = 1;

    int64 sequence_number = 2;

    MetadataUpdateInfo update_info = 3;
}

// Response sent for a failed heart beat.
message HeartbeatNack {}

message ExecuteQueryRequest {
    uuidpb.UUID query_id      = 1 [
                                  (gogoproto.customname) = "QueryID"
                                  ];
    // TODO(philkuz) DEPRECATION planned.
    string query_str = 2;
    pl.carnot.planpb.Plan logical_plan = 3;
}
